controllers/progressivesync_controller.go

```diff
@@ -332,6 +332,28 @@ func (r *ProgressiveSyncReconciler) updateStatusWithRetry(ctx context.Context, p
 	return retryErr
 }
 
+func (r *ProgressiveSyncReconciler) addSyncedAtAnnotation(ctx context.Context, app argov1alpha1.Application, stageName string) error {
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
This ends up being very similar to https://github.com/Skyscanner/applicationset-progressive-sync/blob/main/controllers/progressivesync_controller.go#L318. Do we think there is an opportunity for some refactoring here?

Reply:
This is manipulating argocd apps though, I think it is fine like this, i.e it is fine to build on top of RetryOnConflict.

I would argue though that this is not just adding the annotation it is also updating it if the stage name has changed, so maybe it should be called `setSyncedAtAnnotation` instead? Also, we are missing the comment above the method.

Reply:
Alternatively, maybe this can be made a part of syncApp method instead.

Reply:
I am more in favour of keeping it as a separate function as it will allow for easier unit tests. 

The question for refactoring - there are similarities but also they change and update different elements. The only part that I can see being repeated is when we retrieve the latest app object. So I can take this part out in a separate function, what do you think?

For the function name, yes, I agree, I will update it.

Reply:
Good points - agree with renaming it but keeping the separate function.

Reply:
Sounds like this thread should be resolved :-)

Reply:
Oh, yes, sorry, resolving now. 🙂